test/Microsoft.Extensions.Internal.Test/ObjectMethodExecutorTest.cs

```diff
@@ -0,0 +1,390 @@
+            {
+                parameter.value = "HelloWorld";
+                return Task.FromResult<TestObject>(parameter);
+            }
+
+            public string EchoWithDefaultAttributes(
+                [DefaultValue("hello")] string input1,
+                [DefaultValue(true)] bool input2,
+                [DefaultValue(10)] int input3)
+            {
+                return input1;
+            }
+
+            public string EchoWithDefaultValues(
+                string input1 = "hello",
+                bool input2 = true,
+                int input3 = 20)
+            {
+                return input1;
+            }
+
+            public string EchoWithDefaultValuesAndAttributes(
+                [DefaultValue("Hi")] string input1 = "hello",
+                [DefaultValue(10)] int input3 = 20)
+            {
+                return input1;
+            }
+
+            public string EchoWithNoDefaultAttributesAndValues(
+                string input1,
+                int input2,
+                bool input3,
+                TestObject input4)
+            {
+                return input1;
+            }
+
+            public TestAwaitable<TestObject> CustomAwaitableOfReferenceTypeAsync(
+                string input1,
+                int input2)
+            {
+                return new TestAwaitable<TestObject>(new TestObject
+                {
+                    value = $"{input1} {input2}"
+                });
+            }
+
+            public TestAwaitable<int> CustomAwaitableOfValueTypeAsync(
+                int input1,
+                int input2)
+            {
+                return new TestAwaitable<int>(input1 + input2);
+            }
+        }
+
+        public class TestAwaitable<T>
+        {
+            private T _result;
+            private bool _isCompleted;
+            private List<Action> _onCompletedCallbacks = new List<Action>();
+
+            public TestAwaitable(T result)
+            {
+                _result = result;
+
+                // Simulate a brief delay before completion
+                ThreadPool.QueueUserWorkItem(_ =>
+                {
+                    Thread.Sleep(100);
+                    SetCompleted();
+                });
+            }
+
+            private void SetCompleted()
+            {
+                _isCompleted = true;
+
+                foreach (var callback in _onCompletedCallbacks)
+                {
+                    callback();
+                }
+            }
+
+            public TestAwaiter GetAwaiter()
+            {
+                return new TestAwaiter(this);
+            }
+
+            public struct TestAwaiter : INotifyCompletion
+            {
+                private TestAwaitable<T> _owner;
+
+                public TestAwaiter(TestAwaitable<T> owner) : this()
+                {
+                    _owner = owner;
+                }
+
+                public bool IsCompleted { get; }
```

Comment:
Doesn't affect these tests, but really this should return the value of `_owner._isCompleted`.